docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
+* all values are propagated to a role through `r_` prefixed variables only
+* all values that are set inside a role are read outside of a role through `r_` prefixed variables only
+* all available `r_` prefixed variables are documented and checked before a main role task is run
+* a table of `r_` prefixed variables in README.md is automatically generated
+
+### Terminilogy
+
+**Role variables**: All `r_` prefixed variables read/set inside a role
+
+**Local variables**: All `l_` prefixed variables set inside a role
+
+### Role properties
+
+**Solid API**: all variables that a role caller can set need to be well defined.
+It means each such variable must have the following information attached:
+* name
+* description
+* default value (if the variable is defined)
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
```

Comment:
@michaelgugino, @mtnbikenc @ingvagabund, for variable passing I liked the approach from the ansible roles for fun and profit video.  
```
# call the role
- include_role:
    name: openshift_node
  vars:
    r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args | default('UNDEFINED'}}"
```
Then inside of the role we check for `UNDEFINED` and do the appropriate action.
```
- set_fact:
    r_openshift_node_kubelet_args: "{{ __r_openshift_node_kubelet_args }}"
  when:  r_openshift_node_kube_args == 'UNDEFINED'
```
This approach gives us the most flexibility by handling `undefined` inventory vars and defaulting to the role's variables.  This was mentioned in the talk at [ansiblefest](https://www.ansible.com/ansible-roles-by-acquia) @8:45 by Jeff as a good way to pass variables in and maintain default variables as well.

This localizes all of the special knowledge of the variables to our role instead of the calling playbook.  This does introduce yet another `__` prefix.  Thoughts on this approach?